### PR TITLE
Fixed regression with the ggv pointer persisting when hands are up

### PR DESCRIPTION
**Pull request by RogPodge**
## Overview
Fixes regression where the gaze cursor could still show up with articulated hands present. Also fixed edge case where gaze cursor would remain on the screen if hands were spawned while near an interactable 

Old broken behavior
GazeCursorOld

New behavior
GazeCursorFix

## Changes
- Fixes: #7742

### CODE REVIEW

**Comment by wiwei** [comment]
@RogPodge can you retarget this to the 2.4 stabilization branch?

**Comment by wiwei** [comment]
Is there a test case that we have that could prevent us from regressing this in the future? I see that the change (https://github.com/microsoft/MixedRealityToolkit-Unity/pull/7736/files) did have a few test case modifications, would be interesting to see if we could have something that would prevent this going forward

File: Assets/MRTK/Tests/PlayModeTests/UserInputSimulationTest.cs [hunk 1]
```diff
@@ -81,7 +80,48 @@ public IEnumerator HandsFreeInteractionTest()
...
-        public IEnumerator ArticulatedHandGestureTest()
+        public IEnumerator InputSimulationArtictulatedHandNearGrabbable()
+        {
+            var iss = PlayModeTestUtilities.GetInputSimulationService();
+            TestUtilities.PlayspaceToOriginLookingForward();
+            yield return null;
+
+            // No hands, default cursor should be visible
+            Assert.IsTrue(CoreServices.InputSystem.GazeProvider.GazeCursor.IsVisible, "Head gaze cursor should be visible");
+            //KeyInputSystem.StopKeyInputSimulation();
```
Comment (julenka) [inline Assets/MRTK/Tests/PlayModeTests/UserInputSimulationTest.cs:91]: is this intentionally commented out?

File: Assets/MRTK/Tests/PlayModeTests/UserInputSimulationTest.cs [hunk 2]
```diff
@@ -81,7 +80,48 @@ public IEnumerator HandsFreeInteractionTest()
...
+            yield return null;
+
+            // Grab pointer is near grabbable
+            TestHand hand = new TestHand(Handedness.Right);
+            var grabPointer = hand.GetPointer<SpherePointer>();
+            Assert.IsTrue(grabPointer.IsNearObject, "Grab pointer should be near a grabbable");
+
+            yield return PlayModeTestUtilities.WaitForInputSystemUpdate();
+
+            // Head cursor invisible when grab pointer is near grabbable
```
Comment (julenka) [inline Assets/MRTK/Tests/PlayModeTests/UserInputSimulationTest.cs:119]: Would be good to make sure that hand grabbabel pointer is enabled here (so you can grab stuff). Also would be good to check that hand rays are visible when hand is not near grabbable. So we should have:

1. hand rays are enabled when not near grabbable
2. gaze cursor not visible.
3. When near grabbable, rays are off and the gaze cursor not visible.

**Comment by keveleigh** [comment]
Looks like this branch needs to be rebased on top of the stabilization branch, since it's bringing in some unrelated changes from the dev branch

```
git reset --hard upstream/prerelease/2.4.0_stabilization
git cherry-pick da92626e2435559a789b5edeb05695af9a3248ed
git cherry-pick c4c6b6dbcdb7275de84dd0741be00f5bf21959ad
git push -f
```

File: Assets/MRTK/Services/InputSystem/FocusProvider.cs [hunk 3]
```diff
@@ -1157,7 +1157,8 @@ private void ReconcilePointers()
                 {
                     if (pointerData.Pointer is IMixedRealityNearPointer)
```
Comment (keveleigh) [inline Assets/MRTK/Services/InputSystem/FocusProvider.cs:1158]: I think this should be able to be updated to the following, and you can remove the `var nearPointer = (IMixedRealityNearPointer)pointerData.Pointer;` line, so we aren't checking type / casting twice.


```suggestion
                    if (pointerData.Pointer is IMixedRealityNearPointer nearPointer)
```

File: Assets/MRTK/Tests/PlayModeTests/UserInputSimulationTest.cs [hunk 4]
```diff
@@ -81,7 +80,53 @@ public IEnumerator HandsFreeInteractionTest()
         }
 
         [UnityTest]
-        public IEnumerator ArticulatedHandGestureTest()
+        public IEnumerator InputSimulationArtictulatedHandNearGrabbable()
```
Comment (keveleigh) [inline Assets/MRTK/Tests/PlayModeTests/UserInputSimulationTest.cs:83]:
```suggestion
        public IEnumerator InputSimulationArticulatedHandNearGrabbable()
```

File: Assets/MRTK/Services/InputSystem/FocusProvider.cs [hunk 5]
```diff
@@ -1155,7 +1155,7 @@ private void ReconcilePointers()
 
                 foreach (var pointerData in pointers.Values)
                 {
-                    if (pointerData.Pointer is IMixedRealityNearPointer)
+                    if (pointerData.Pointer is IMixedRealityNearPointer nearPointer)
                     {
                         var nearPointer = (IMixedRealityNearPointer)pointerData.Pointer;
```
Comment (RogPodge) [inline Assets/MRTK/Services/InputSystem/FocusProvider.cs:1160]:
```suggestion
```

File: Assets/MRTK/Services/InputSystem/FocusProvider.cs [hunk 6]
```diff
@@ -1155,9 +1155,9 @@ private void ReconcilePointers()
 
                 foreach (var pointerData in pointers.Values)
                 {
-                    if (pointerData.Pointer is IMixedRealityNearPointer)
+                    if (pointerData.Pointer is IMixedRealityNearPointer nearPointer)
                     {
-                        if (pointerData.Pointer.IsInteractionEnabled)
+                        if (nearPointer.IsInteractionEnabled || nearPointer.IsNearObject)
```
Comment (keveleigh) [inline Assets/MRTK/Services/InputSystem/FocusProvider.cs:1160]: What does this change do? I think most (all of the ones I looked at, at least) of our near pointers use `IsNearObject` as part of their `IsInteractionEnabled` checks. Do we care if a near pointer is near an object but not able to interact with it due to some other part of the `IsInteractionEnabled` check? My gut says we wouldn't, but maybe I'm missing something

Comment (RogPodge) [reply Assets/MRTK/Services/InputSystem/FocusProvider.cs:1160]: the near pointer's we have are the sphere pointer and the poke pointer. IsNearObject is used to toggle between near and far interactions, which is slightly different from IsInteractionEnabled. As such, the NumNearPointersActive count should go up if the IsNearObject flag is true. 

The bug this addresses is when an articulated hand spawns near an object, but not quite in the range to trigger IsInteractionEnabled. In these instances, the gazecursor would remain on the screen, even though the cursor won't be on the screen if you spawn the hand far away from the object and then move closer to it

**Pull request merged**